Documentation for linking Stacks (#36408)
hashicorp/terraform: website/docs/language/stacks/reference/tfdeploy.mdx

| :---- | :---- | :---- | :---- |
| `inputs` | A mapping of Stack variable names for this deployment.  The keys in this map must correspond to the names of variables defined in the Stack. The values must be valid HCL literals meeting the type constraint of those variables. | map | Required |

> ### Reference

Author: All the changes in this file are adding "Reference" headers above examples (so no actual example changes) as per Adam's request!

## `locals` block configuration

A local value assigns a name to an expression, so you can use the name multiple times within your Stack configuration instead of repeating the expression. The `locals` block works exactly as it does in traditional Terraform configurations. Learn more about [the `locals` block](/terraform/language/values/locals).

## `publish_output` block configuration

> -> **Note**: To use the `publish_output` block requires at least Terraform version `terraform_1.10.0-alpha20241009` or higher. We recommend downloading the [latest version of Terraform](https://releases.hashicorp.com/terraform/) to use the most up-to-date functionality.

Reviewer: ```suggestion
This block requires at Terraform version `terraform_1.10.0-alpha20241009` or higher. Download the [latest version of Terraform](https://releases.hashicorp.com/terraform/) to use the most up-to-date functionality.
```

> To export information from one Stack to another, declare the `publish_output` block in your deployment configuration to export data that another Stack can intake.

Reviewer: ```suggestion
Specifies a value to export from your Stack. Refer to [Pass data from one Stack to another](//terraform/language/stacks/deploy/link-stacks) for additional information. 
```
The reference should just describe the behavior.

Author: I'm going to split the difference a bit to give folks some context here:

```suggestion
Specifies a value to export from your current Stack, which other Stacks in the same project can consume. Declare one `publish_output` block for each value to export your Stack. 
```

> The `publish_output` block is similar to the `output` block in the Terraform configuration language, but the `publish_output` block declares values available at the Stack level. Add a `publish_output` block for each value you want to output from your current Stack.

Reviewer: ```suggestion
The `publish_output` block  declares values available at the Stack level. You must declare one `publish_output` block for each value you want to output from your current Stack.
```
Comparing elements in this way can be frustrating for new users.

### Complete configuration

When every field is defined, a `publish_output` block has the following form:

<CodeBlockConfig hideClipboard>

```hcl
publish_output "output_name" {
  description = "Description of the purpose of this output"
  value       = deployment.deployment_name.some_value
}
```

</CodeBlockConfig>

### Specification

This section provides details about the fields you can configure in the output block.

| Field | Description | Type | Required |
| :---- | :---- | :---- | :---- |
| `value` | The value to output.  | any | Required |
| `description` | A human-friendly description for the output. | string | Optional |

> For example, you could output the VPC ID from your networking deployment, making it available to other Stacks to input.
> 
> <CodeBlockConfig filename="network.tfdeploy.hcl">
> 
> ```hcl
> # Network Stack's deployment configuration
> 
> publish_output "vpc_id" {
>   description = "The networking Stack's VPC's ID."
>   # Referencing a particular deployment can be helpful for sharing information
>   # across your Stacks.
>   value       = deployment.network.vpc_id
> }
> ```
> 
> </CodeBlockConfig>
> 
> This example uses the `publish_output` block to expose the `vpc_id` output from the `network` deployment of this Stack. Another Stack in the same project could then intake your `vpc_id` output by declaring an `upstream_input` block. 
> 
> To learn more about passing information between Stacks, refer to [Pass data from one Stack to another](/terraform/language/stacks/deploy/link-stacks).

Reviewer: ````suggestion
For example, you could output the VPC ID from your networking deployment, making it available to other Stacks to input.

<CodeBlockConfig filename="network.tfdeploy.hcl">

```hcl
# Network Stack's deployment configuration

publish_output "vpc_id" {
  description = "The networking Stack's VPC's ID."
  # Referencing a particular deployment can be helpful for sharing information
  # across your Stacks.
  value       = deployment.network.vpc_id
}
```

</CodeBlockConfig>

This example uses the `publish_output` block to expose the `vpc_id` output from the `network` deployment of this Stack. Another Stack in the same project could then intake your `vpc_id` output by declaring an `upstream_input` block. 

To learn more about passing information between Stacks, refer to [Pass data from one Stack to another](/terraform/language/stacks/deploy/link-stacks).
````
For consistency, place all of the examples in the configuration reference in a section called "Reference" after the "Specification" section.

Author: This is the reference for a whole type of file (these blocks are unique to the `.tfdeploy.hcl` file), so I'll add `Reference` headers under each block for each one's example!

## `upstream_input` block configuration 

> -> **Note**: The `upstream_input` block requires at least Terraform version `terraform_1.10.0-alpha20241009` or higher. We recommend downloading the [latest version of Terraform](https://releases.hashicorp.com/terraform/) to use the most up-to-date functionality.
> 
> To read outputs from other Stacks in your project, declare an `upstream_input` block for each Stack you want to reference. If the Stack's output values change, HCP Terraform automatically triggers runs for any Stacks that depend on those outputs.  
> 
> We refer to the Stack that declares an `upstream_input` block as the downstream Stack. An upstream Stack publishes outputs using the `publish_output` block, and a downstream Stack consumes those outputs. For example, if Stack B relies on outputs from Stack A, Stack B is the downstream Stack.

Reviewer: ```suggestion
Requires at Terraform version `terraform_1.10.0-alpha20241009` or higher. Download the [latest version of Terraform](https://releases.hashicorp.com/terraform/) to use the most up-to-date functionality.

Specifies a source of data to use in the Stack configuration. Declare an `upstream_input` block for each Stack you want to reference. To learn more about passing information between Stacks, refer to [Pass data from one Stack to another](/terraform/language/stacks/deploy/link-stacks). 
```

Author: Tweaking to make more technically accurate:

```suggestion
The `upstream_input` block requires at least Terraform version `terraform_1.10.0-alpha20241009` or higher. Download [latest version of Terraform](https://releases.hashicorp.com/terraform/) to use the most up-to-date functionality.

The `upstream_input` block specifies another Stack in the same project from which to consume outputs. Declare an `upstream_input` block for each Stack you want to reference. If an output from a upstream Stack changes, HCP Terraform automatically triggers runs for any Stacks that depend on those outputs. 

To learn more about passing information between Stacks, refer to [Pass data from one Stack to another](/terraform/language/stacks/deploy/link-stacks). 
```
